ndk/src/system_properties.rs

```diff
@@ -0,0 +1,327 @@
+//! Bindings for [System Properties]
+//!
+//! [System Properties]: https://source.android.com/docs/core/architecture/configuration/add-system-properties
+
+#[cfg(feature = "api-level-26")]
+use std::mem::MaybeUninit;
+use std::{
+    ffi::{c_char, c_void, CStr, CString, FromBytesWithNulError, FromVecWithNulError},
+    fmt,
+    io::Error,
+    num::NonZeroU32,
+    ptr::NonNull,
+    str::{FromStr, Utf8Error},
+    time::Duration,
+};
+
+use thiserror::Error;
+
+use crate::utils::{abort_on_panic, status_to_io_result};
+
+/// Possible failures returned by [`get_raw()`].
+#[derive(Debug, Error)]
+pub enum GetRawError {
+    #[error("Property is missing or empty")]
+    MissingOrEmpty,
+    #[error(transparent)]
+    NulError(#[from] FromVecWithNulError),
+    #[error(transparent)]
+    Io(#[from] Error),
+}
+
+/// Internal helper to deduplicate the implementation between [`get_raw()`] and
+/// [`Property::read_raw()`].
+fn process_owned(get: impl FnOnce(*mut c_char) -> i32) -> Result<CString, GetRawError> {
+    // Pre-allocate a `Vec` which we can move to the user with the result
+    let mut value = Vec::with_capacity(ffi::PROP_VALUE_MAX as usize);
+    let ret = get(value.as_mut_ptr());
+    match ret {
+        0 => Err(GetRawError::MissingOrEmpty),
+        ..0 => Err(Error::from_raw_os_error(-ret).into()),
+        1.. => {
+            // TODO: This "smart" implementation leaves the user with a 92-byte allocation since
+            // set_len() currently doesn't shrink.  Any such operation would likely reallocate,
+            // making this have no advantage over the stack-local variant that allocates after the
+            // fact.
+            unsafe { value.set_len(ret as usize + 1) }
+            Ok(CString::from_vec_with_nul(value)?)
+        }
+    }
+}
+
+/// Returns the property value as an owned [`CString`] with possibly invalid UTF-8 [but no interor
+/// NULs].  The maximum length can be up to 92 ([`ffi::PROP_VALUE_MAX`]) including NUL terminator.
+///
+/// [but no interor NULs]: GetRawError::NulError
+///
+/// See [`get()`] for a more convenient API that validates this string for UTF-8 and directly parses
+/// it into a [`FromStr`]-compatible type.
+///
+/// # Deprecation
+/// Deprecated since Android O (API level 26), use [`Property::find()`] with
+/// [`Property::read_callback()`] instead which does not have a limit on `value` nor `name` length.
+#[doc(alias = "__system_property_get")]
+pub fn get_raw(name: &CStr) -> Result<CString, GetRawError> {
+    process_owned(|value| unsafe { ffi::__system_property_get(name.as_ptr(), value) })
+}
+
+/// Possible failures returned by [`get()`].
+#[derive(Debug, Error)]
+pub enum GetError<T> {
```

Comment:
Yikes, our MSRV is too old and this Rust version doesn't emit the `Debug` implementation behind a `T: Debug` boundary. Instead requiring us to force a `T: Debug` and `<T as FromStr>>::Err: Debug` bound throughout the codebase or not have `.unwrap()` capabilities at all (which requires the `Error` type to be `Debug`able):

https://github.com/rust-mobile/ndk/actions/runs/13476131028/job/37655556180?pr=495#step:6:108

Reply:
This also makes it inconvenient/impossible to use the type inside the error wrapper for https://github.com/rust-mobile/ndk/pull/479.

Reply:
As it turns out this wasn't a broken bound on `Debug` but on the generated `impl`s from `thiserror::Error`. This was fixed a couple years ago but our MSRV CI test intentionally downgrades our crates: we just had to bump the minimum bound a little to `1.0.29`. As a bonus we not only no longer